### PR TITLE
[function] Add ROUND() & TRUNCATE() functions

### LINKED ISSUE #2563: [functions] more functions working with number
**Summary**

From MySQL docs in https://dev.mysql.com/doc/refman/8.0/en/numeric-functions.html

Tasks list:

If you are interested in one or more tasks in this issue, you can comment like: `/assignme abs functions` to make this subtask assigned to you, then I'll mark that as wip. 
- [x] add a development doc <how to develop functions in databend>  to guide developers.
- [ ] document of the below functions, refer: https://dev.mysql.com/doc/refman/8.0/en/mathematical-functions.html
- [x] %, MOD	Modulo operator
- [x] *	Multiplication operator
- [x] +	Addition operator
- [x] -	Minus operator
- [x] -	Change the sign of the argument
- [x] /	Division operator
- [x] ABS()	Return the absolute value [#2596] 
- [x] ACOS()	Return the arc cosine [#2888]
- [x] ASIN()	Return the arc sine [#2888]
- [x] ATAN()	Return the arc tangent [#2888]
- [x] ATAN2(), ATAN()	Return the arc tangent of the two arguments [#2888]
- [x] CEIL()	Return the smallest integer value not less than the argument[#2759]
- [x] CEILING()	Return the smallest integer value not less than the argument[#2759]
- [wip] CONV()	Convert numbers between different number bases
- [x] COS()	Return the cosine [#2610]
- [x] COT()	Return the cotangent [#2610]
- [x] CRC32()	Compute a cyclic redundancy check value [#2616]
- [x] DEGREES()	Convert radians to degrees [#2643]
- [x] DIV	Integer division 
- [x] EXP()	Raise to the power of e [#2847]
- [x] FLOOR()	Return the largest integer value not greater than the argument[#2759]
- [x] LN()	Return the natural logarithm of the argument [#2733]
- [x] LOG()	Return the natural logarithm of the first argument [#2669 ]
- [x] LOG10()	Return the base-10 logarithm of the argument [#2733 ]
- [x] LOG2()	Return the base-2 logarithm of the argument [#2733 ]
- [x] MOD()	Return the remainder [#2850 ]
- [x] PI()	Return the value of pi [#2584]
- [x] POW()	Return the argument raised to the specified power [#2949]
- [x] POWER()	Return the argument raised to the specified power [#2949]
- [x] RADIANS()	Return argument converted to radians [#2643]
- [WIP] RAND()	Return a random floating-point value
- [x] ROUND()	Round the argument [#2923]
- [x] SIGN()	Return the sign of the argument [#2906]
- [x] SIN()	Return the sine of the argument [#2610]
- [x] SQRT()	Return the square root of the argument [#2905 ]
- [x] TAN()	Return the tangent of the argument [#2610]
- [x] TRUNCATE()	Truncate to specified number of decimal places [#2923]

### PR DESCRIPTION
**Pull request by mshauneu**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

- Add ROUND() function
- Add TRUNCATE() function

## Changelog

- New Feature

## Related Issues

Related #2563

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/functions/tests/it/scalars/maths/round.rs [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+                ),
+                DataColumnWithField::new(
+                    DataColumn::Constant(DataValue::Int64(Some(35)), 1),
+                    DataField::new("d", DataType::Int64, false),
+                ),
+            ],
+            input_rows: 2,
+            func: RoundNumberFunction::try_create("round")?,
+            expect: Series::new([
+                0.123456789012345678901234567890,
```
Comment (mshauneu) [inline common/functions/tests/it/scalars/maths/round.rs:317]: this test works fine but when I run it from cmd sql I see some truncation in output:

```txt
mysql> SELECT ROUND(0.12345678901234567890123456789012345, 35);
+--------------------------------+
| ROUND(0.12345678901234568, 35) |
+--------------------------------+
|            0.12345678901234568 |
+--------------------------------+
```

Maybe it is related to just output...?

File: website/databend/docs/sqlstatement/numeric-functions/round.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
++-----------------+
+
+SELECT ROUND(23.298, -1);
++----------------------+
+| ROUND(23.298, (- 1)) |
++----------------------+
+|                   20 |
++----------------------+
+
+SELECT ROUND(0.12345678901234567890123456789012345, 35);
```
Comment (mshauneu) [inline website/databend/docs/sqlstatement/numeric-functions/round.md:73]: Unlike real mysql engine we truncate our numbers due to the f64 precision limitation. To support exactly mysql behavior maybe extra DataType::Decimal (https://docs.rs/rust_decimal/1.17.0/rust_decimal/) need to be introduced.

File: website/databend/docs/sqlstatement/numeric-functions/round.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
++--------------------------------+
+| ROUND(0.12345678901234568, 35) |
++--------------------------------+
+|            0.12345678901234568 |
++--------------------------------+
```
Comment (mshauneu) [inline website/databend/docs/sqlstatement/numeric-functions/round.md:78]: Unlike real mysql engine we truncate our numbers due to the f64 precision limitation. To support exactly mysql behavior maybe extra DataType::Decimal (https://docs.rs/rust_decimal/1.17.0/rust_decimal/) need to be introduced.

Comment (sundy-li) [reply website/databend/docs/sqlstatement/numeric-functions/round.md:78]: Does not matter, we can add Decimal datatype later.

File: common/functions/src/scalars/maths/round.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+                                    x_series
+                                        .f64()?
+                                        .iter()
+                                        .map(|x| x.map(|x| (self.rounding_func)(*x))),
+                                )
+                            }
+                            Ok(v) => match v.cmp(&0) {
+                                Ordering::Greater => {
+                                    let z = 10_f64.powi(if v > 30 { 30 } else { v as i32 });
+                                    DFFloat64Array::new_from_opt_iter(x_series.f64()?.iter().map(
```
Comment (sundy-li) [inline common/functions/src/scalars/maths/round.rs:98]: ArrayApply is better than `new_from_opt_iter`.

Comment (mshauneu) [reply common/functions/src/scalars/maths/round.rs]: Thanks, change it.

File: website/databend/docs/sqlstatement/numeric-functions/round.md (hunk 3 above)
Comment (mshauneu) [reply website/databend/docs/sqlstatement/numeric-functions/round.md:78]: Got it.

File: common/functions/src/scalars/maths/round.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                    let opt_iter = x_series
+                        .f64()?
+                        .into_iter()
+                        .zip(d_series.i64()?.into_iter())
+                        .map(|(x, d)| match (x, d) {
+                            (Some(x), Some(d)) => round(x, d),
+                            _ => None,
+                        });
+
+                    DFFloat64Array::new_from_opt_iter(opt_iter)
```
Comment (sundy-li) [inline common/functions/src/scalars/maths/round.rs:128]: validity did not take into account.

try `new_from_iter_validity()`

File: common/functions/src/scalars/maths/round.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                            }),
+                            Ordering::Equal => Some((self.rounding_func)(*x)),
+                        }
+                    };
+
+                    let opt_iter = x_series
+                        .f64()?
+                        .into_iter()
+                        .zip(d_series.i64()?.into_iter())
+                        .map(|(x, d)| match (x, d) {
```
Comment (sundy-li) [inline common/functions/src/scalars/maths/round.rs:123]: validity can be combined at last.

try `new_from_iter_validity`

Comment (sundy-li) [reply common/functions/src/scalars/maths/round.rs:123]: After #2944, we can use binary function.

Comment (mshauneu) [reply common/functions/src/scalars/maths/round.rs]: 👍 
switched to binary

**Comment by Junnplus** [comment]
Some case error are not expected.
```
mysql> select number, round(pi(), number) from numbers(10) order by number;
+--------+---------------------+
| number | round(pi(), number) |
+--------+---------------------+
|      0 |                   3 |
|      1 |                 3.1 |
|      2 |                3.14 |
|      3 |               3.142 |
|      4 |              3.1416 |
|      5 |             3.14159 |
|      6 |            3.141593 |
|      7 |           3.1415927 |
|      8 |           3.1415927 |
|      9 |           3.1415927 |
+--------+---------------------+
10 rows in set (0.00 sec)
Read 10 rows, 80 B in 0.002 sec., 5.58 thousand rows/sec., 44.66 KB/sec.

mysql> select round(pi(), 8);
+--------------------+
| round(pi(), 8)     |
+--------------------+
| 3.1415926499999998 |
+--------------------+
1 row in set (0.00 sec)
Read 1 rows, 1 B in 0.000 sec., 3.08 thousand rows/sec., 3.08 KB/sec.

mysql> select round(pi(), 7);
+----------------+
| round(pi(), 7) |
+----------------+
|      3.1415927 |
+----------------+
1 row in set (0.00 sec)
Read 1 rows, 1 B in 0.000 sec., 2.49 thousand rows/sec., 2.49 KB/sec.

mysql> select round(pi(), 9);
+----------------+
| round(pi(), 9) |
+----------------+
|    3.141592654 |
+----------------+
1 row in set (0.00 sec)
Read 1 rows, 1 B in 0.000 sec., 2.51 thousand rows/sec., 2.51 KB/sec.
```

**Comment by mshauneu** [comment]
8 is magic number :-)
https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=59e1ffe00888929f0177cf9e4126bf5c
```
3
3.1
3.14
3.142
3.1416
3.14159
3.141593
3.1415927
3.1415926499999998
3.141592654
3.1415926536
3.14159265359
3.14159265359
3.1415926535898
3.14159265358979
3.141592653589793
3.141592653589793
3.141592653589793
3.141592653589793
3.141592653589793
```
will take care tmrw

**Comment by sundy-li** [comment]
There is a round function in cpp named `nearbyint` and `nearbyintf`
http://www.cplusplus.com/reference/cmath/nearbyint/

Rust seems to have [it](https://github.com/rust-lang/rust/blob/master/library/core/src/intrinsics.rs#L1463-L1467) too.

**Comment by sundy-li** [comment]
Seems `nearbyint` is unstable, so forget above comment.

But why not just use: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=bced371991b57f06ff6ab9441462c09d

```
3
3.1
3.14
3.142
3.1416
3.14159
3.141593
3.1415927
3.14159265
3.141592654
3.1415926536
3.14159265359
3.14159265359
3.1415926535898
3.14159265358979
3.141592653589793
3.141592653589793
3.141592653589793
3.141592653589793
3.141592653589793
```

**Comment by Junnplus** [comment]
@mshauneu Another problem is that we should check that the first column is constant, not just the second column.

```
mysql> select number, round(pi(), number) from numbers(13) order by number;

+--------+---------------------+
| number | round(pi(), number) |
+--------+---------------------+
|      0 |                   3 |
|      1 |                 3.1 |
|      2 |                3.14 |
|      3 |               3.142 |
|      4 |              3.1416 |
|      5 |             3.14159 |
|      6 |            3.141593 |
|      7 |           3.1415927 |
|      8 |           3.1415927 |
|      9 |           3.1415927 |
|     10 |           3.1415927 |
|     11 |           3.1415927 |
|     12 |           3.1415927 |
+--------+---------------------+
13 rows in set (0.01 sec)
```

**Comment by mshauneu** [comment]
@sundy-li - i started from approach you posted, then I thought that it could overflow result. Overflowing is last what user expect from rounding function...

**Comment by sundy-li** [comment]
Wow, can you post an example that may cause results overflowing?

**Comment by mshauneu** [comment]
@sundy-li https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=91403ef6c03c3210eca48dfa7f0fef2b

**Comment by mshauneu** [comment]
@Junnplus @sundy-li 

- on precision: 
maybe it is ok to accept overflow at least we will have decimal type support 
(https://www.reddit.com/r/rust/comments/j2mhre/how_to_do_arithmetic_with_f64_numbers_without/ )

- on any column could be a const:
great point, implement this, please see truc test cases
(https://github.com/datafuselabs/databend/blob/2f1da027df550cd9011e340919baab586a491443/common/functions/tests/it/scalars/maths/round.rs#L399)
```
mysql -h0.0.0.0 -P3307 -uroot -e 'select number, round(pi(), number) from numbers(16) order by number'
+--------+---------------------+
| number | round(pi(), number) |
+--------+---------------------+
|      0 |                   3 |
|      1 |                 3.1 |
|      2 |                3.14 |
|      3 |               3.142 |
|      4 |              3.1416 |
|      5 |             3.14159 |
|      6 |            3.141593 |
|      7 |           3.1415927 |
|      8 |          3.14159265 |
|      9 |         3.141592654 |
|     10 |        3.1415926536 |
|     11 |       3.14159265359 |
|     12 |       3.14159265359 |
|     13 |     3.1415926535898 |
|     14 |    3.14159265358979 |
|     15 |   3.141592653589793 |
+--------+---------------------+
```

**Comment by sundy-li** [comment]
> maybe it is ok to accept overflow at least we will have decimal type support

Yes, I agree to accept overflow for f64 currently. Yet even with decimal type, there would be  overflow range of decimal, it's ok to output the result as `inf` or `null` if overflow happens.

**Review by sundy-li — APPROVED** [review]

**Pull request merged**